libraries/chain/fork_database.cpp

```diff
@@ -341,114 +402,155 @@ namespace eosio::chain {
+      if( bs_accessor_t::is_valid(**candidate) ) {
          head = *candidate;
       }
    }
 
-   template<class bsp>
-   void fork_database_t<bsp>::add( const bsp& n, bool ignore_duplicate ) {
+   template<class BSP>
+   void fork_database_t<BSP>::add( const bsp_t& n, mark_valid_t mark_valid, ignore_duplicate_t ignore_duplicate ) {
       std::lock_guard g( my->mtx );
-      my->add_impl( n, ignore_duplicate, false,
+      my->add_impl( n, mark_valid, ignore_duplicate, false,
                     []( block_timestamp_type timestamp,
                         const flat_set<digest_type>& cur_features,
                         const vector<digest_type>& new_features )
                     {}
       );
    }
 
-   template<class bsp>
-   bsp fork_database_t<bsp>::root() const {
+   template<class BSP>
+   bool fork_database_t<BSP>::has_root() const {
+      return !!my->root;
+   }
+
+   template<class BSP>
+   BSP fork_database_t<BSP>::root() const {
       std::lock_guard g( my->mtx );
       return my->root;
    }
 
-   template<class bsp>
-   bsp fork_database_t<bsp>::head() const {
+   template<class BSP>
+   BSP fork_database_t<BSP>::head() const {
       std::lock_guard g( my->mtx );
       return my->head;
    }
 
-   template<class bsp>
-   bsp fork_database_t<bsp>::pending_head() const {
+   template<class BSP>
+   BSP fork_database_t<BSP>::pending_head() const {
       std::lock_guard g( my->mtx );
-      const auto& indx = my->index.template get<by_lib_block_num>();
+      const auto& indx = my->index.template get<by_best_branch>();
 
       auto itr = indx.lower_bound( false );
-      if( itr != indx.end() && !(*itr)->is_valid() ) {
+      if( itr != indx.end() && !fork_database_impl<BSP>::bs_accessor_t::is_valid(**itr) ) {
          if( first_preferred( **itr, *my->head ) )
             return *itr;
       }
 
       return my->head;
    }
 
-   template <class bsp>
-   fork_database_t<bsp>::branch_type
-   fork_database_t<bsp>::fetch_branch(const block_id_type& h, uint32_t trim_after_block_num) const {
+   template <class BSP>
+   fork_database_t<BSP>::branch_t
+   fork_database_t<BSP>::fetch_branch(const block_id_type& h, uint32_t trim_after_block_num) const {
       std::lock_guard g(my->mtx);
       return my->fetch_branch_impl(h, trim_after_block_num);
    }
 
-   template <class bsp>
-   fork_database_t<bsp>::branch_type
-   fork_database_impl<bsp>::fetch_branch_impl(const block_id_type& h, uint32_t trim_after_block_num) const {
-      branch_type result;
+   template <class BSP>
+   fork_database_t<BSP>::branch_t
+   fork_database_impl<BSP>::fetch_branch_impl(const block_id_type& h, uint32_t trim_after_block_num) const {
+      branch_t result;
+      result.reserve(index.size());
+      for (auto i = index.find(h); i != index.end(); i = index.find((*i)->previous())) {
+         if ((*i)->block_num() <= trim_after_block_num)
+            result.push_back(*i);
```

Comment:
```suggestion
         if ((*i)->block_num() > trim_after_block_num)
            break;
         result.push_back(*i);
```

Reply:
This doesn't work. See new test case which fails with this implementation.

Reply:
I'm baffled! Do you know why my version doesn't work or should I look into it?

Reply:
Oh never mind I was again misled by the name of the parameter.